tools/lxdclient/client_image.go

```diff
@@ -92,56 +97,107 @@ func (p *progressContext) copyProgress(progress string) {
 		}
-		logger.Infof("found image from %s for %s = %s",
-			source.URL(), series, target)
-		forwarder := stringforwarder.New(copyProgressHandler)
-		defer func() {
-			dropCount := forwarder.Stop()
-			logger.Debugf("dropped %d progress messages", dropCount)
-		}()
-		adapter := &progressContext{
-			logger:  logger,
-			level:   loggo.INFO,
-			context: fmt.Sprintf("copying image for %s from %s: %%s", name, source.URL()),
-			forward: forwarder.Forward,
+		if lastErr = err; lastErr == nil {
+			break
+		}
+	}
+	return imageName, lastErr
+}
+
+func (i *imageClient) ensureImage(
+	series, imageName string,
+	aliases []string,
+	source remoteClient,
+	copyProgressHandler func(string),
+) error {
+	// Look for an image with any of the aliases.
+	var alias, target string
+	for _, alias = range aliases {
+		if target = source.GetAlias(alias); target != "" {
+			break
 		}
-		err = source.CopyImage(series, i.raw, []string{name}, adapter.copyProgress)
-		return errors.Annotatef(err, "unable to get LXD image for %s", name)
 	}
-	return lastErr
+	if target == "" {
+		// TODO(jam) Add a test that we skip sources that don't
+		// have what we are looking for
+		logger.Infof("no image for %s found in %s", imageName, source.URL())
+		return errors.NotFoundf("image for %s in %s", imageName, source.URL())
+	}
+
+	logger.Infof("found image from %s for %s = %s", source.URL(), imageName, target)
+	forwarder := stringforwarder.New(copyProgressHandler)
+	defer func() {
+		dropCount := forwarder.Stop()
+		logger.Debugf("dropped %d progress messages", dropCount)
+	}()
+	adapter := &progressContext{
+		logger:  logger,
+		level:   loggo.INFO,
+		context: fmt.Sprintf("copying image for %s from %s: %%s", imageName, source.URL()),
+		forward: forwarder.Forward,
+	}
+	err := source.CopyImage(alias, i.raw, []string{imageName}, adapter.copyProgress)
+	return errors.Annotatef(err, "unable to get LXD image for %s", imageName)
+}
+
+// seriesLocalAlias returns the alias to assign to images for the
+// specified series. The alias is juju-specific, to support the
+// user supplying a customised image (e.g. CentOS with cloud-init).
+func seriesLocalAlias(series, arch string) string {
+	return fmt.Sprintf("juju/%s/%s", series, arch)
 }
 
-// A common place to compute image names (aliases) based on the series
-func (i imageClient) ImageNameForSeries(series string) string {
-	// TODO(jam) Do we need 'ubuntu' in there? We only need it if "series"
-	// would collide, but all our supported series are disjoint
-	return fmt.Sprintf("ubuntu-%s", series)
+// seriesRemoteAliases returns the aliases to look for in remotes.
+func seriesRemoteAliases(series, arch string) ([]string, error) {
+	seriesOS, err := jujuseries.GetOSFromSeries(series)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	switch seriesOS {
+	case os.Ubuntu:
+		return []string{path.Join(series, arch)}, nil
+	case os.CentOS:
+		if series == "centos7" && arch == "amd64" {
+			return []string{"centos/7/amd64"}, nil
```

Comment:
Do we have any idea what a correct value for this actually is?
This definitely feels like the sort of thing where we need more than one "image-metadata-url", because we're probably going to need a different upstream source for CentOS images than 'cloud-images.ubuntu.com'